docs/architecture/adr-005-modular-server.md

```diff
@@ -0,0 +1,166 @@
+[others](https://github.com/cosmos/cosmos-sdk/blob/f9dc082059d63423f96315ff913a8745c3178a7d/server/api/server.go#L81-L119) implement them as separate functions as the ones above. The lack of a standard approach for
+these steps results in difficulty for extensibility as each of these functions are individually
+called by the `start` command after checking if the service is enabled or not by the configuration.
+
+### App configuration extensibility
+
+Each service relies on the configuration options defined on `app.toml`. Thus, in order to extend
+the services provided by the server, the SDK must handle custom TOML files provided by each
+application.
+
+## Decision
+
+The proposed approach standardizes the server and service setup process. 
+To accomplish this a new `Service` and `Server` interface will be introduced based on the 3 steps outlined above:
+
+```go
+type Service interface {
+    RegisterRoutes() bool
+    Start(config.ServerConfig) error
+    Stop() error
+}
+
+type Server interface {
+    GetServices() []Service
+    RegisterServices() error
+    Start(config.ServerConfig) error
+    Stop() error
+}
+```
+
+### Configuration
+
+Since the enablement of a service depends on the application configuration, an additional
+`ServerConfig` interface is required so that the current configuration utility functions are still
+applicable for the SDK and extensible concrete configurations:
+
+```go
+// ServerConfig extends the SDK default configuration TOML
+type ServerConfig interface {
+    // The SDK config defines the 5 configurations used by default: Base, Telemetry, API, gRPC and State Sync. 
+    GetSDKConfig() *Config
+}
+```
+
+### Start Command
+
+```go
+func startInProcess(ctx *Context, clientCtx client.Context, appCreator types.AppCreator) error {
+  // ...
+  app := appCreator(ctx.Logger, db, traceWriter, ctx.Viper)
+
+  // handle application-specific configuration
+  cfg := app.GetConfig(ctx.Viper)
+  // ...
+
+  // 1. Create application-specific server
+  server := app.NewServer(clientCtx, ctx.Logger, cfg)
+
+  // 2. Register routes from each service 
+  // NOTE: routes are registered regardless if the service is enabled or not
+  if err := server.RegisterServices(); err != nil {
+    return err
+  }
+
+  // 3. Start each of the services.
+  // NOTE: each service must check if it's enabled via the configuration
+  if err := server.Start(cfg); err != nil {
+    return err
+  }
+
+  defer func() {
+    // ...
+    // stop all the services
+    if err := server.Stop(); err != nil {
+      ctx.Logger.Error("failed to stop server", "error", err)
+    }
+    // ...
+  }
+  // ...
+}
+```
+
+## Consequences
+
+
+### Positive
+
+- Standardize all the services provided by the SDK
+- Modularize the server so that it allows for custom application-specific services.
+
+### Negative
```

Comment:
I would add
 
- more modularity adds more of a maintenance burden. 
- Can lead to user mistakes in registration and/or server configuration

Love the approach being taken but too much configuration like this can lead to many user mistakes. I believe the sdk should be opinionated in this category to help prevent user errores